src/server/pkg/storage/chunk/writer.go

```diff
@@ -322,272 +272,188 @@ func (w *worker) upload(path string, chunk []byte) error {
-		if len(w.bufAnnotations) > 0 {
-			lastA := w.bufAnnotations[len(w.bufAnnotations)-1]
-			if lastA.drs[0].DataRef().ChunkInfo.Chunk.Hash != dr.DataRef().ChunkInfo.Chunk.Hash {
-				if err := w.flushDataReaders(); err != nil {
-					return err
-				}
+		// Create the cross chunk references.
+		// We keep track of the previous referenced chunk to prevent duplicate create
+		// reference calls.
+		for _, dataRef := range a.RefDataRefs {
+			refChunk := dataRef.ChunkInfo.Chunk.Hash
+			if prevRefChunk == refChunk {
+				continue
 			}
-		}
-		// Join annotation with current buffered annotations, then buffer the current data reader.
-		w.bufAnnotations = joinAnnotations(w.bufAnnotations, copyAnnotation(a))
-		lastA := w.bufAnnotations[len(w.bufAnnotations)-1]
-		lastA.drs = append(lastA.drs, dr)
-		w.bufSize += dr.Len()
-		// Cheap copy if full chunk is buffered.
-		if w.bufSize == dr.DataRef().ChunkInfo.SizeBytes {
-			for _, a := range w.bufAnnotations {
-				a.NextDataRef = a.drs[0].DataRef()
+			if err := w.gcC.CreateReference(w.ctx, &gc.Reference{
+				Sourcetype: "chunk",
+				Source:     path.Join(prefix, chunkRef.ChunkInfo.Chunk.Hash),
+				Chunk:      path.Join(prefix, refChunk),
+			}); err != nil {
+				return err
 			}
-			annotations := w.bufAnnotations
-			w.fs = append(w.fs, func() error {
-				return w.f(annotations)
-			})
-			// (bryce) need to handle less than WindowSize chunk.
-			w.lastCopiedAnnotation = w.bufAnnotations[len(w.bufAnnotations)-1]
-			w.bufAnnotations = nil
-			w.bufSize = 0
+			prevRefChunk = refChunk
 		}
 	}
 	return nil
 }
 
-func (w *worker) atSplit() bool {
-	return !w.first && w.numBytesRolled() == 0
+func sizeOfTags(tags []*Tag) int64 {
+	var size int64
+	for _, tag := range tags {
+		size += tag.SizeBytes
+	}
+	return size
 }
 
-func (w *worker) rollDataReader(a *Annotation, dr *DataReader) error {
-	if err := dr.Iterate(func(tag *Tag, r io.Reader) error {
-		_, err := io.Copy(a.buf, r)
-		if err != nil {
-			return err
+func (w *Writer) executeFunc(annotations []*Annotation, prevChan, nextChan chan struct{}) error {
+	// Wait for the previous chunk to be processed before executing callback.
+	if prevChan != nil {
+		select {
+		case <-prevChan:
+		case <-w.ctx.Done():
+			return w.ctx.Err()
 		}
-		a.tags = append(a.tags, tag)
-		return nil
-	}); err != nil {
+	}
+	if err := w.f(annotations); err != nil {
 		return err
 	}
-	return w.roll(a)
+	if nextChan != nil {
+		close(nextChan)
+	}
+	return nil
 }
 
-func (w *worker) flushDataReaders() error {
-	for _, a := range w.bufAnnotations {
-		for _, dr := range a.drs {
-			if err := w.rollDataReader(copyAnnotation(a), dr); err != nil {
-				return err
-			}
-		}
+// Copy copies data from a data reader to the writer.
+// The copy will either be by reading the referenced data, or just
+// copying the data reference (cheap copy).
+func (w *Writer) Copy(dr *DataReader) error {
+	if w.done() {
+		return w.eg.Wait()
```

Comment:
This line is a little unexpected to me, my understanding is that `w.done` indicates that the context has been cancelled, when that happens I normally expect the function to return immediately with an error. Is the idea that the functions passed to the `eg` will fail with the relevant error, and we don't want to return until they've run so we wait on them and get the error from there? Is it possible for `w.done()` to return true and this code not return an error?

Reply:
When a context is cancelled aren't you supposed to return `ctx.Err()` basically indicating that nothing went wrong, and you stopped early to satisfy the constraints of the caller? If something happens in cleanup definitely log it but I don't think it should necessarily make it back to the caller as the cause.

Reply:
`w.done` is checking a context derived from the passed in context. This derived context will get canceled if any of the goroutines error. So, yeah, we use wait to collect the error and ensure we don't leak goroutines. This also accounts for the passed in context getting canceled. I do think you could have `w.done` be true and not return an error when no goroutines are running and the parent context gets canceled. I am going to change this to be more comprehensive. I am thinking something like this:
```
func (w *Writer) Write(data []byte) (int, error) {
	if err := w.maybeDone(func() error {
		if err := w.flushDataReaders(); err != nil {
			return err
		}
		w.roll(data)
		return nil
	}); err != nil {
		return 0, err
	}
	return len(data), nil
}

func (w *Writer) maybeDone(func() error) (retErr error) {
	if w.err != nil {
		return w.err
	}
	defer func() {
		if retErr != nil {
			w.err = retErr
			w.cancel()
		}
	}
	select {
	case <-w.ctx.Done():
		if err := w.eg.Wait(); err != nil {
			return err
		}
		return w.ctx.Err()
	default:
	}
	return f()
}
```
The context hierarchy would be (passed in ctx - > cancelable ctx (`w.cancel`) -> errgroup context) Basically, we will capture the first error (there is a caveat to this) and store it for return at each of the exported functions. This accounts for the passed in context getting canceled, a goroutine erroring (main or error group), and the writer getting closed (it will call `w.cancel`). This approach is not air tight as far as capturing the first error is concerned because an error could occur in an error group goroutine before an error in the main goroutine callback, but the main goroutine callback error would be recorded. I think this is fine particularly because handling that requires me to use wait groups instead of error groups which adds a fair amount of complexity for pretty low value. What do you guys think?

Reply:
could you have a `func (w *Writer) done() error` method that uses a `sync.Once` to do cleanup and returns the error from the context.  Then start all the exported methods with
```
if err := w.done(); err != nil {
    return err
}
```

Reply:
I think anytime you're using a `ctx` there's a race condition between the request finishing and the cancel function being called, I think it's implicitly part of the semantics that it's possible to call `cancel` and not receive an error, or receive a different error, because the request was about to return anyways. I think this race is fairly superficial (and also basically unavoidable) the important thing is that the request shouldn't block for a long time after `cancel` is called. And if it didn't succeed it should return an error of some sort.